features/mbedtls/targets/TARGET_Silicon_Labs/crypto_aes.c

```diff
@@ -0,0 +1,545 @@
+    CRYPTO_DataWriteUnaligned(&device->DATA0, (uint8_t *)input);
+    CORE_EXIT_CRITICAL();
+
+    if ( mode == MBEDTLS_AES_ENCRYPT ) {
+        device->CMD = CRYPTO_CMD_INSTR_AESENC;
+    } else {
+        device->CMD = CRYPTO_CMD_INSTR_AESDEC;
+    }
+    while ((device->STATUS & CRYPTO_STATUS_INSTRRUNNING) != 0);
+
+    CORE_ENTER_CRITICAL();
+    CRYPTO_DataReadUnaligned(&device->DATA0, (uint8_t *)output);
+    CORE_EXIT_CRITICAL();
+
+    crypto_management_release(device);
+
+    return ret;
+}
+
+#if defined(MBEDTLS_CIPHER_MODE_CBC)
+
+/*
+ * AES-CBC buffer encryption/decryption
+ */
+int mbedtls_aes_crypt_cbc( mbedtls_aes_context *ctx,
+                           int mode,
+                           size_t length,
+                           unsigned char iv[16],
+                           const unsigned char *input,
+                           unsigned char *output )
+{
+    int ret = 0;
+    CORE_DECLARE_IRQ_STATE;
+    size_t processed = 0;
+
+    if( ctx == NULL || input == NULL || output == NULL || iv == NULL ) {
+        return ( MBEDTLS_ERR_AES_INVALID_INPUT_LENGTH );
+    }
+
+    /* Input length must be a multiple of 16 bytes which is the AES block
+       length. */
+    if( length & 0xf ) {
+        return( MBEDTLS_ERR_AES_INVALID_INPUT_LENGTH );
+    }
+
+    if ( ctx->keybits != 128UL && ctx->keybits != 256UL) {
+        return MBEDTLS_ERR_AES_INVALID_KEY_LENGTH;
+    }
+
+    CRYPTO_TypeDef *device = crypto_management_acquire();
+    device->WAC = 0;
+    device->CTRL = 0;
+
+    CORE_ENTER_CRITICAL();
+    CRYPTO_KeyBufWrite(device, (uint32_t*)ctx->key, (ctx->keybits == 128UL) ? cryptoKey128Bits : cryptoKey256Bits);
+    if ( mode == MBEDTLS_AES_ENCRYPT ) {
+        CRYPTO_DataWriteUnaligned(&device->DATA0, (uint8_t *)iv);
+    } else {
+        CRYPTO_DataWriteUnaligned(&device->DATA2, (uint8_t *)iv);
+    }
+    CORE_EXIT_CRITICAL();
+
+    while ( processed < length ) {
+        if ( mode == MBEDTLS_AES_ENCRYPT ) {
+            CORE_ENTER_CRITICAL();
+            CRYPTO_DataWriteUnaligned(&device->DATA0XOR, (uint8_t *)(&input[processed]));
+            device->CMD = CRYPTO_CMD_INSTR_AESENC;
+            CRYPTO_DataReadUnaligned(&device->DATA0, (uint8_t *)(&output[processed]));
+            CORE_EXIT_CRITICAL();
+        } else {
+            /* Decrypt input block, XOR IV to decrypted text, set ciphertext as next IV */
+            CORE_ENTER_CRITICAL();
+            CRYPTO_DataWriteUnaligned(&device->DATA0, (uint8_t *)(&input[processed]));
+            CRYPTO_EXECUTE_4( device,
+                              CRYPTO_CMD_INSTR_DATA0TODATA1,
+                              CRYPTO_CMD_INSTR_AESDEC,
+                              CRYPTO_CMD_INSTR_DATA2TODATA0XOR,
+                              CRYPTO_CMD_INSTR_DATA1TODATA2);
+            CRYPTO_DataReadUnaligned(&device->DATA0, (uint8_t *)(&output[processed]));
+            CORE_EXIT_CRITICAL();
+        }
+        processed += 16;
```

Comment:
It can process a multiple of the block size, but then it needs to know the entire size upfront. Additionally, it cannot be preempted during the entire operation when chaining multiple blocks, and we need the preemption support (including low preemption latency) for wireless stacks.